scylla/src/transport/connection.rs

```diff
@@ -488,11 +488,18 @@ async fn connect_with_source_port(
     addr: SocketAddr,
     source_port: u16,
 ) -> Result<TcpStream, std::io::Error> {
-    // TODO: handle ipv6?
-    let socket = TcpSocket::new_v4()?;
-    let source_addr = IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0));
-    socket.bind(SocketAddr::new(source_addr, source_port))?;
-    Ok(socket.connect(addr).await?)
+    match addr {
+        SocketAddr::V4(_) => {
+            let socket = TcpSocket::new_v4()?;
+            socket.bind(format!("0.0.0.0:{}", source_port).parse().unwrap())?;
```

Comment:
This line:

- Allocates a new string on the heap,
- Writes `0.0.0.0:` to it and appends to it the `source_port` converted to decimal,
- Parses the complete address.

While I don't think this function is on a critical path, there is no need to do all of that in order to initialize a `SocketAddr` structure. Using `SocketAddr::new` is much cheaper:

```rust
let source_addr = IpAddr::V4(Ipv4Addr::new(0, 0, 0, 0));
socket.bind(SocketAddr::new(source_addr, source_port))?;
```

Similar change should be done to the IPv6 case, too.